File: Algrithms-Level-1/day-6/s1.c
int lengthOfLongestSubstring(char *s)
{
    int map[128] = {0}, l = 0;
    char *end = s, *start = s;

    while (*end)
    {
        if (map[*end])
        {
            l = (end - start > l) ? end - start : l;
            while (*start != *end)
            {
                map[*start] = 0;
                start++;
            }
            end++;
            start++;
        }
        else
        {
            map[*end] = 1;
            end++;
        }
    }
    l = (end - start > l) ? end - start : l;
    return l;
}

File: Algrithms-Level-1/day-6/s1.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0

        found = {}
        start = result = 0

        for i in range(len(s)):
            if s[i] in found and start <= found[s[i]]:
                start = found[s[i]] + 1
            else:
                result = max(result, i - start + 1)
            found[s[i]] = i

        return result
